src_c/geometry.c

```diff
@@ -7,6 +7,108 @@
 #endif /* ~__AVX2__ */
 
 #define PYGAMEAPI_GEOMETRY_NUMSLOTS 21
+static PyObject *
+pg_raycast(PyObject *self, PyObject *args, PyObject *kwargs)
+{
+    PyObject *startpoint;
+    PyObject *endpoint = NULL;
+    PyObject **farr;
+    PyObject **collisions;
+    Py_ssize_t loop;
+    double angle;
+    double max_dist;
+    double target_pos[2] = {0, 0};
+
+    static char *keywords[] = {"start_pos", "collisions", "endpoint", "angle", "max", NULL};
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "OO|Odd",
+        keywords, &startpoint, &collisions, &endpoint, &angle, &max_dist))
+    {
+        return NULL;
+    }
```

Comment:
As i said in discord i'm not completely fine with kwargs here. There should be more thought put into this. 
### 1
You have two parameters for start  and end positions, where one is way at the start and the other way at the back. We could be smarter than this and have the function accept as first argument either a sequence of `(start_pos, end_pos)` (meaning we have a tuple containing two tuples) or a single value (tuple) containing the ray origin. The function should be resposible to check the type that's passed and act accordingly. In this way we reduce the param. number by one.

### 2
The `start_pos` and `collisions` parameters should not be kwargs, as you always need at least these two, and they should always be at the start of the parameter list passed for grammatical coherence and clarity. So you should remove them from the `static char *keywords[]` array.

### 3
We could further remove parameters if we consider that a ray can be represented in two ways (at least for what we have right now) which are:
1. A Sequence containing two tuples which represent the start and end position of the ray. It's obviout that in this case the ray already has a max_length that's iplicit to the Line length. `((x1, y1), (x2, y2))`
2. We could also imagine a ray as a **Sequence of start_pos, angle, max_distance**. In that case we would simply represent the ray as a Tuple of `(start_pos, angle, max_distance)` or `((x, y), angle, max_distance)`

Given these two options it's apparent how we wouldn't need to pass the extra `endpoint`,  `angle`, `max` parameters as we could simply have a single parameter that represents both: a `ray` parameter if you wish, that's automatically calculated and converted to a usable object.

So combining all that i said we would have something like this:

```C
    static char *keywords[] = {"ray", "collisions", NULL};
    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "OO", keywords, &ray, &collisions))
    {
        return NULL;
    }
```
As you can see now we can't have kwargs anymore as it would be superflous. At this point you could revert it back to FASTCALL and have the function take care of the first argument.

Reply:
I don't like the idea of having tuples inside tuples, would this be a better API design instead? This will use FASTCALL instead.
```python
@overload
def raycast(originpos, endpos, colliders): ...
@overload
def raycast(originpos, angle, max_dist, colliders): ...
```